### ISSUE TITLE
Builder Pattern API

### ISSUE THREAD

**Issue by jquesada2016** [sycamore-rs/sycamore#268]
### Proposal

I have been contemplating what a builder pattern API, as mentioned in #265, might look like. I personally like the idea of having both API's, and having the interop, as macros have their place which can be much easier to read, but the primary advantages, in my opinion to a builder pattern is that, currently, formatting and type hints don't work with the current tooling (rust-analyzer), and certain more complicated logic is not possible at the moment, without further syntax extensions.

Also, from a discord comment, @lukechu10 mentioned the aim of eventually supporting React Native, and possibly other rendering targets. This to me, presents two different routes that could be taken:

1. Having a rendering-agnostic API
2. Having a rendering-specific API

The first option could be something as follows (for HTML):

```rust
button();
div();
a();
/* etc */

```

Whereas the agnostic approach could be something like:

```rust
node("button");
node("div");
node("a");

```

These features could be held behind a feature flag, such as `html`, `react-native`, `native-script`, etc. for the render-specific API, or just `builder-api` for the agnostic API, and prefixed with `experimental:` until an API is stabilized.

### Main required features

In no particular order. Also note that all of the mentioned functions would be implemented for the builder struct and take either `&mut self` or `self` and return their respective `Self` type, but have been omitted for brevity..

- (Un)setting Attributes
    - A further distinction could be made (for memory and performance reasons) between static and dynamic attributes, such as:

        ```rust
        fn set_attribute(name: &str, value: impl Into<JsValue>);

        // The Option type is important here, as it can be used to automatically remove
        // the attribute when required.
        fn set_dyn_attribute(name: &str, value: StateHandle<Option<impl Into<JsValue>>>);

        // Also note that remove_attribute is not needed here, since it's handled in the
        // dynamic case and in static cases not needed.
        ```

- (Un)setting properties
    - Same as above
- (Un)setting classes
    - Once again, I recommend having a distinction between static and dynamic classes.

        ```rust
        fn add_class(class: &str);
        fn add_dyn_class(class: StateHandle<Option<&str>>);
        ```

- Add/remove children
    - For setting children, we have some choices to make. We can accept `Vec<GenericNode>`, or `Template`, thought the latter is preferable. The question lies on how to handle dynamic children.

        ```rust
        fn add_children(children: Template<G>);

        // Perhaps dynamic children could be handled as follows
        fn add_dyn_children(children: StateHandle<Template<G>>);
        ```

- Special shortcuts
    - `fn id(id: &str)`
    - `fn style(name: &str, value: &str)` or perhaps `fn add_styles<K, V>(HashMap<K, V>) where K: Hash + AsRef<str>, V: AsRef<str>` or something similar. along with their dynamic `Signal` accepting counterparts.
    - And possible others I've missed.
- Event listeners

    ```rust
    fn add_event_listener(event: &str /* or typed Event */, handler: EventHandler);

    // When listen is false, the event handler won't be discarded, just not trigger until
    // listen is true again.
    fn add_dyn_event_listener(event: &str, handler: EventHandler, listen: StateHandle<bool>);
    ```

- Refs

    ```rust
    fn ref(ref: NodeRef);
    ```

- Binding
    - Each respective supported bindable property could be made, such as `value`, `checked`, and optional element specific binding can be made, such as `group`.

        ```rust
        fn bind_value(value: Signal<String>);
        fn bind_group(group: Signal<String>);
        /* etc */
        ```

    - Another very important note, especially with Web Components, binding to arbitrary attributes and properties is a must have, as a lot of components, it's the only way to get notified on changes witch aren't emitted as custom events.

    ```rust
    fn bind_custom_attribute(name: &str, Signal<impl Deserialize>);
    fn bind_custom_property(name: &str, Signal<impl Deserialize>);
    ```

- Build
    - Finally, we need to build the `node` into a `Template`.

### Important note

The above API would currently require using Template for instantiating components, but it could be mentioned in the docs that if macros would like to be completely omitted, then one could use `Component::__create_component(props)`.

Did I miss anything? Can anything be improved? Please don't be shy. The more the critique, the better we can make Sycamore, together.

**Comment by arctic-hen7** [comment]
I think the render-agnostic approach looks excellent (and supporting mobile native platforms in future would be amazing), and this API looks brilliant! Great job!

**Comment by jquesada2016** [comment]
I started a POC for the renderer-specific version. I'll see how far I get today. I plan on creating a working POC for both.

**Comment by lukechu10** [comment]
Thanks for the write up!

> * For setting children, we have some choices to make. We can accept `Vec<GenericNode>`, or `Template`, thought the latter is preferable. The question lies on how to handle dynamic children.

I was thinking of using constant generics here (Sycamore's MSRV is 1.53.0 which supports it already) to make it as short as possible. For dynamic children, we can just use `Template::new_dyn` which is what the `template!` macro expands to now anyways.

Also instead of using `StateHandle<_>` for the dynamic variants, I would prefer using just a `Fn` so that it makes it easier to pass dynamic expressions. In fact, we can make a new trait so that both `Fn`, `StateHandle` and `Signal` all work without any more effort.

**Comment by jquesada2016** [comment]
> I was thinking of using constant generics here

What do you mean with using const generics in this context?

> Also instead of using StateHandle<_> for the dynamic variants, I would prefer using just a Fn so that it makes it easier to pass dynamic expressions. In fact, we can make a new trait so that both Fn, StateHandle and Signal all work without any more effort.


This actually sounds like a great idea! If you could provide a quick example on the trait signature and/or in use, that would be great.

If you could provide a short example for both of these points, I'll include it in my POC which I will try to get done either today or by latest tomorrow.

**Comment by jquesada2016** [comment]
@lukechu10, I know you mentioned you would like to implement a trait for accepting both `StateHandle` and `Signal` for the API, but are you sure that's the best choice? I'm only thinking that stating the intention of a function call accepting only `StateHandle` tells the consumer that their data won't be changed internally, and vice-versa for `Signal`, stating that their data _can_ change.

**Comment by lukechu10** [comment]
> What do you mean with using const generics in this context?

I meant being able to do something like:
```rust
div()
    .with([
        a()
        button()
        etc...
    ])
```

And so the `.with` function would need to be generic over the length of the array. This is slightly shorter than `vec![...]` or even `&[...]` but it shouldn't affect binary size nor compile times because it would get inlined most likely anyways.

**Comment by jquesada2016** [comment]
ahh I see. That should be fine, though the `.with` name seems a bit odd.

I've added all the basic functionality I think is necessary (probably missing things) for the agnostic (`node("div")`) API. Now we can work on the render specific API (along with the names of the functions, if they need to change, or be added).
This should be as easy as just making an attribute macro for an `HtmlTag` enum that will create all of the named element functions.

**Comment by jquesada2016** [comment]
Just added all HTML tags.

Can now be used like
```rust
div()
    .add_only_child(
        button()
            .add_text("Click me")
            .add_event_listener("click", || { web_sys::console::log_1(&"Button clicked!".into()) })
    )
    .build();
```

**Comment by lukechu10** [comment]
Just merged #269 so I'm closing this now.
